### PR TITLE
[GraphTrainer] Introduce CPU offload pass for activation memory savings

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #3015

Adds a graph-level CPU offload pass that asynchronously transfers activation
tensors to CPU during forward and reloads them during backward, reducing peak
GPU memory usage.

- Define offload/reload/wait_tensor op insertions using upstream
  torch._functorch._activation_offloading.offload_ops
- Use CheckpointPolicy.MUST_CPU_OFFLOAD from torch.utils.checkpoint
- Use meta.get("custom", {}).get(_MODULE_FQN) as layer boundary markers and seq_nr
  for forward/backward node classification
- Skip view ops, small tensors (<4KB), collectives, and the last layer
- No-op when no nodes are tagged
- Handle symbolic FakeTensors (DSv3 MoE) gracefully
- Integrate into apply_default_graph_passes (make_fx path) and
  AVAILABLE_JOINT_PASSES registry (AOT path)
- 7 unit tests covering tagging, op insertion, edge cases, and no-op behavior

### CODE REVIEW

**Comment by mlazos** [comment]
@claude can you analyze this for any additional cleanups or general improvements (e.g. dead code, edge cases in the pass handling or stylistic improvements)

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+        aten.expand,
+        aten.permute,
+        aten.as_strided,
+        aten.alias,
+        aten.split,
+    }
+)
+
+# Collective/wait ops whose outputs should not be offloaded
+_COLLECTIVE_OPS = frozenset(
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: maybe dumb q: are there any ops in torch.ops._c10d_functional that should _not_ be considered as collective ops?

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+    {
+        torch.ops._c10d_functional.all_reduce,
+        torch.ops._c10d_functional.all_gather_into_tensor,
+        torch.ops._c10d_functional.reduce_scatter_tensor,
+        torch.ops._c10d_functional.all_to_all_single,
+        torch.ops._c10d_functional.wait_tensor,
+    }
+)
+
+_MIN_OFFLOAD_BYTES = 4096  # 4 KB minimum to justify offload overhead
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/cpu_offload.py:75]: how did we arrive at this number?

File: torchtitan/experiments/graph_trainer/passes.py [hunk 3]
```diff
@@ -54,6 +56,10 @@
 )
 from torchtitan.tools.logging import logger
 
+if TYPE_CHECKING:
+    from torchtitan.experiments.graph_trainer.configs import GraphTrainerCompileConfig
+    from torchtitan.experiments.graph_trainer.trainer import GraphTrainer
+
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/passes.py]: don't need?

File: torchtitan/experiments/graph_trainer/tests/test_passes.py [hunk 4]
```diff
@@ -909,10 +915,319 @@ def test_chained_identity_slices(self):
         self.assertEqual(self._count_slice_nodes(gm), 0)
 
 
-# Import TestCustomCodegenPass so it's discoverable when running this file
-from torchtitan.experiments.graph_trainer.tests.test_custom_codegen import (  # noqa: F401
-    TestCustomCodegenPass,
-)
+class TestCpuOffloadPass(TestCase):
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/tests/test_passes.py]: Follow the pattern of TestCustomCodegenPass, have it in another file and import it here

File: torchtitan/experiments/graph_trainer/tests/test_passes.py [hunk 5]
```diff
@@ -909,10 +915,319 @@ def test_chained_identity_slices(self):
...
+    """Unit tests for the CPU offload pass on synthetic FX graphs.
+
+    Tests use hand-built graphs that simulate the joint fwd+bwd structure
+    produced by make_fx, with ``seq_nr`` for forward/backward classification
+    and ``ac_region_id`` for layer boundaries.
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/tests/test_passes.py]: FYI, I have removed ac_region_id in 
https://github.com/pytorch/torchtitan/pull/3050

One can directly get layer_id from node.meta["module_fqn"]

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+    last_layer_id = max(all_layer_ids) if len(all_layer_ids) > 1 else None
+
+    tagged = 0
+    total_bytes = 0
+
+    for node in gm.graph.nodes:
+        if node not in forward_nodes:
+            continue
+        layer_id = node.meta.get("custom", {}).get(_AC_REGION_ID)
+        if layer_id is None or layer_id == last_layer_id:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: add a comment here, 

it's an arbitrary heuristic here to not offloading the last layer, since it would be reload in backward soon.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+        gm: The GraphModule containing the full fwd+bwd graph.
+        example_inputs: Example inputs (unused, required by graph pass interface).
+
+    Returns:
+        The transformed GraphModule with offload/reload ops inserted.
+    """
+    # Lazy import to avoid circular dependency (passes.py imports cpu_offload.py)
+    from torchtitan.experiments.graph_trainer.passes import tlparse_log_graph_pass
+
+    tlparse_log_graph_pass(gm, graph_name="cpu_offload_before")
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: don't need here anymore

it's done in the apply_graph_passes

enabled by --compile.debug_graph_passes

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+    # interleaves forward and backward nodes (unlike manually built graphs
+    # where they are cleanly separated), so the simple "defer to next layer's
+    # forward" / "prefetch to preceding layer's backward" strategy from the
+    # reference implementation needs adaptation. For now, offload/reload ops
+    # are placed right next to their producers/consumers, which still saves
+    # memory but without async transfer overlap.
+
+    gm.graph.lint()
+    gm.recompile()
+    tlparse_log_graph_pass(gm, graph_name="cpu_offload_after")
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: same here, not needed anymore

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+    # 2. Build position index for ordering queries (only used for pre-existing
+    # nodes; newly inserted nodes never appear in bwd_users).
+    node_to_index: dict[Node, int] = {n: i for i, n in enumerate(gm.graph.nodes)}
+
+    # 3. Insert offload/reload/wait_tensor ops
+    total_bytes = 0
+
+    for node, bwd_users in offloadable:
+        val = node.meta.get("val")
+        if val is None:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: assert val is not None.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+
+    Returns:
+        The transformed GraphModule with offload/reload ops inserted.
+    """
+    # Lazy import to avoid circular dependency (passes.py imports cpu_offload.py)
+    from torchtitan.experiments.graph_trainer.passes import tlparse_log_graph_pass
+
+    tlparse_log_graph_pass(gm, graph_name="cpu_offload_before")
+
+    _, backward_nodes = _classify_forward_backward(gm)
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: it's slightly expensive to rerun _classify_forward_backward here. 

Since you are only using backward nodes in 303
you can rewrite it as

bwd_users = [u for u in node.users if _is_backward_node(u)]

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+    # are placed right next to their producers/consumers, which still saves
+    # memory but without async transfer overlap.
+
+    gm.graph.lint()
+    gm.recompile()
+    tlparse_log_graph_pass(gm, graph_name="cpu_offload_after")
+    logger.info(
+        f"CPU offload: offloaded {len(offloadable)} tensors "
+        f"({total_bytes / 1024 / 1024:.2f} MB)"
+    )
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py:362]: let's also logger.debug here for all offloaded nodes

File: torchtitan/experiments/graph_trainer/passes.py [hunk 12]
```diff
@@ -84,6 +90,7 @@ def compile_time_passes(
         remove_detach_pass,
         remove_identity_view_pass,
         remove_identity_slice_pass,
+        cpu_offload_pass,
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/passes.py]: let's comment out cpu_offload_pass here for now, as we will need to sort out the common tagging API between offload and sac.

**Review by SherlockNoMad — COMMENTED** [review]
Thanks. The major comment is to remove dependency on ac_region_id. 
Should be good to land once comments are addressed.

File: torchtitan/experiments/graph_trainer/passes.py (hunk 3 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/passes.py]: Hmm ruff was flagging this on my local runs ..

Comment (mlazos) [reply torchtitan/experiments/graph_trainer/passes.py]: nvm doesn't look like it was used

**Comment by mlazos** [comment]
Next steps: actually attempt overlap, and provide API for marking for offload and https://github.com/pytorch/torchtitan/pull/3015#discussion_r3126381549

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+        return False
+    return _has_offloadable_tensor(node)
+
+
+# ============================================================
+# Forward/backward node classification for make_fx traced graphs
+# ============================================================
+
+
+def _classify_forward_backward(
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/cpu_offload.py:156]: looks like multiple passes may care about this distinction -- we should consider making this a shared utility in follow-up. cc @yiming0416

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+    forward_nodes: set[Node] = set()
+    backward_nodes: set[Node] = set()
+
+    for node in gm.graph.nodes:
+        if node.op not in ("call_function", "get_attr"):
+            continue
+        if node.meta.get("autograd_backward", False):
+            backward_nodes.add(node)
+        else:
+            forward_nodes.add(node)
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/cpu_offload.py:176]: Will all-gathers added by SAC pass for unsharding in backwards have this tag?

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+
+    This is a reference implementation that offloads all offloadable
+    activations. In practice, users should tag activations via annotations
+    or graph passes for more fine-grained control over which activations
+    are offloaded.
+
+    Sets ``node.meta["recompute"] = CheckpointPolicy.MUST_CPU_OFFLOAD`` on
+    eligible nodes. Eligibility requires:
+      1. Node is a forward node (determined via seq_nr)
+      2. Node has ``meta["custom"]["ac_region_id"]`` (layer boundary marker)
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: Who introduces this ac_region_id marker?

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 13 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py:156]: +1

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 2 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py:75]: it's arbitrary, selected by claude.

File: torchtitan/experiments/graph_trainer/tests/test_passes.py (hunk 5 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/tests/test_passes.py]: I need to actually remove this from the comment lol

File: torchtitan/experiments/graph_trainer/tests/test_cpu_offload.py [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+from torch.testing._internal.common_utils import TestCase
+from torch.utils.checkpoint import CheckpointPolicy
+
+
+class TestCpuOffloadPass(TestCase):
+    """Unit tests for the CPU offload pass on synthetic FX graphs.
+
+    Tests use hand-built graphs that simulate the joint fwd+bwd structure
+    produced by make_fx, with ``seq_nr`` for forward/backward classification
+    and ``ac_region_id`` for layer boundaries.
```
Comment (mlazos) [inline torchtitan/experiments/graph_trainer/tests/test_cpu_offload.py]: remove his

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 15 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: yeah I need to remove this actually, we can directly get layer_id from node.meta["module_fqn"] which I think the tracer populates, but @SherlockNoMad can correct me if I'm wrong

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 1 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: actually no, I'll jus enumerate the namespace

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 14 above)
Comment (mlazos) [reply torchtitan/experiments/graph_trainer/cpu_offload.py:176]: I don't think so, for offloading this shouldn't matter, because they would be excluded by the collective check anyway. This will matter if we want to make the tag infra more general though

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+
+    return forward_nodes, backward_nodes
+
+
+# ============================================================
+# Tagging pass
+# ============================================================
+
+
+def tag_all_offloadable_activations(gm: torch.fx.GraphModule) -> None:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: follow graph pass signature

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 18]
```diff
@@ -0,0 +1,399 @@
...
+
+# Pattern to extract the layer index from a module FQN like "layers.0.attention.wq"
+_LAYER_INDEX_RE = re.compile(r"layers\.(\d+)")
+
+
+def _is_backward_node(node: Node) -> bool:
+    return node.meta.get("autograd_backward", False)
+
+
+def _get_layer_id(node: Node) -> int | None:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: already exist in passes.py

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 19]
```diff
@@ -0,0 +1,399 @@
...
+from torchtitan.experiments.graph_trainer.common_utils import _MODULE_FQN
+from torchtitan.tools.logging import logger
+
+aten = torch.ops.aten
+
+# Pattern to extract the layer index from a module FQN like "layers.0.attention.wq"
+_LAYER_INDEX_RE = re.compile(r"layers\.(\d+)")
+
+
+def _is_backward_node(node: Node) -> bool:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: already exist in passes.py

**Review by SherlockNoMad — APPROVED** [review]

**Comment by mlazos** [comment]
@claude can you take a final look to see if there are any other minor improvements or unused code
